Test/dafny0/Strings.dfy

```diff
@@ -31,12 +36,35 @@ method Main()
   var s, t := M(ch, ch);
   print "ch = ", ch, "\n";
   print "The string is: " + s + "\n";
-  var x, y, z := Escapes();
+  var x, y, z, zz := Escapes();
   print "Escape X: ", x, "\n";
   print "Escape Y: ", y, "\n";
   print "Escape Z: ", z, "\n";
+  print "Escape ZZ: ", zz, "\n";
   var c, d := CharEscapes();
   print "Here is the end" + [c, d] + [' ', ' ', ' '] + [[d]][0] + "   ", d, "\n";
+
+  var x?, y?, z? := WeirdStrings();
+  
+  // Printing these invalid (in UTF-16) strings can lead to at least inconsistent
+  // output across the backends, but they should never crash.
+  // We assert that the invalid state is modelled correctly as well.
+  expect |x?| == 30;
+  expect x?[29] as int == 55296;
+  print "Weird string X: ", x?, "\n";
+  expect |y?| == 30;
+  expect x?[29] as int == 55296;
+  print "Weird string Y: ", y?, "\n";
+  expect |z?| > 2;
+  expect z?[0..2] == ['\ude0e', '\ud83d'];
+  print "Weird string Z: ", z?, "\n";
+
+  var c?, d? := WeirdChars();
+  print "These characters are quite confused: ", c?, ' ', d?, "\n";
+
+  // Ensuring we're precise enough about identifying \u escapes
+  print "I'm afraid you'll find escape quite impossible, \\u007", "\n";
+  print "Luckily I have this nifty gadget from my good friend, \\\u0051", "\n";
```

Comment:
Callback to my favourite test case ever: https://github.com/dafny-lang/dafny/blob/master/Test/expectations/Expect.dfy#L19-L22 :)